packages/loopring_v3/contracts/amm/libamm/AmmData.sol

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../core/iface/ExchangeData.sol";
+import "../../core/iface/IExchangeV3.sol";
+
+
+/// @title AmmData
+library AmmData
+{
+    function LP_TOKEN_BASE() internal pure returns (uint) { return 10 ** 18; }
+    function LP_TOKEN_INITIAL_SUPPLY() internal pure returns (uint) { return 100 * LP_TOKEN_BASE(); }
+    function MAX_AGE_REQUEST_UNTIL_POOL_SHUTDOWN() internal pure returns (uint) { return 7 days; }
+    function MIN_TIME_TO_UNLOCK() internal pure returns (uint) { return 1 days; }
+
+    enum PoolTransactionType
+    {
+        NOOP,
+        JOIN,
+        EXIT
+    }
+
+    struct PoolConfig
+    {
+        address   exchange;
+        string    poolName;
+        uint32    accountID;
+        address[] tokens;
+        uint96[]  weights;
+        uint8     feeBips;
+        string    tokenSymbol;
+    }
+
+    struct PoolJoin
+    {
+        address  owner;
+        bool     fromLayer2;
+        uint     minPoolAmountOut;
+        uint96[] maxAmountsIn;
+        uint32[] storageIDs;
+        uint     validUntil;
+    }
+
+    struct PoolExit
+    {
+        address  owner;
+        bool     toLayer2;
+        uint     poolAmountIn;
+        uint96[] minAmountsOut;
+        uint32[] storageIDs;
+        uint     validUntil;
+    }
+
+    struct PoolTransaction
+    {
+        PoolTransactionType txType;
+        bytes               data;
+        bytes               signature;
+    }
+
+    struct Token
+    {
+        address addr;
+        uint96  weight;
+        uint16  tokenID;
+    }
+
+    struct Context
+    {
+        ExchangeData.Block _block;
+        IExchangeV3        exchange;
+        address            exchangeDepositContract;
+
+        uint     txIdx;
+        bytes32  domainSeperator;
+        bytes32  exchangeDomainSeparator;
+        uint96[] ammActualL2Balances;
+        uint96[] ammExpectedL2Balances;
+        uint     numTransactionsConsumed;
+
+        Token[]  tokens;
+        uint     size; // == token.length;
+
+        uint     poolTokenTotalSupply;
+        uint     poolTokenBase;
+        uint     poolTokenInitialSupply;
+    }
+
+    struct State {
+        // Pool token state variables
+        string name;
+        string symbol;
+        uint   totalSupply;
+
+        mapping(address => uint) balanceOf;
+        mapping(address => mapping(address => uint)) allowance;
+        // Question: should we use the same `nonces` variable?
+        mapping(address => uint) permitNonces; // ERC2612
+
+        // AMM pool state variables
+        IExchangeV3 exchange;
+        uint32      accountID;
+        bytes32     domainSeperator;
```

Comment:
I don't know why the previous PR was dropped, but I committed some small changes there that did some minor improvements. One of them was the type here -> separator: https://github.com/Loopring/protocols/pull/1788/commits/a47c18ef09956349f8993cb4379f2f5448a2c8be

Reply:
Oh I misread, I see you'll still fix the issues in previous PR.

Reply:
yes

Reply:
https://github.com/Loopring/protocols/commit/a47c18ef09956349f8993cb4379f2f5448a2c8be merged